### PR TITLE
feat: v1.9.0 — bats test suite, structured category field, gate extraction, auto-cheap routing, novelty pass

### PR DESCRIPTION
**Pull request by gchaix**
## Summary

- **12 issues resolved** (#67–#78): bug fixes, new features, test infrastructure, and documentation corrections
- **54 bats tests** added covering all deterministic bash helpers (`run-cve-check.sh`, `run-trufflehog.sh`, `evaluate-gates.sh`) and golden orchestration contracts
- **Breaking fixes**: Go CVE `replace` directive ordering (#67), `Write` tool missing from frontmatter (#68), GitHub inline review MCP doc mismatch (#69), TruffleHog wrong input mode (#71)
- **New features**: structured `category` field in `json-findings` (#76), auto-cheap routing for docs/config diffs (#78), novelty pass for repeated findings (#77), functional `/comprehensive-review-help` skill (#73)
- Supersedes v1.8.11 (governance changes included in base branch `feat/v1.8.11-governance-alignment`)

## Changes by issue

| Issue | Type | Description |
|-------|------|-------------|
| #67 | fix | `parse_go_mod` two-pass awk: `require`-before-`replace` now resolves correctly |
| #68 | fix | `"Write"` added to `allowed-tools` frontmatter (enables `--output-file`) |
| #69 | fix | GitHub inline review doc updated: `gh api`, not `mcp__github-pat__*` |
| #70 | fix | PR/MR `body`/`description` fetched in `--pr` mode for all providers |
| #71 | fix | TruffleHog invocation switched from diff-file to per-file (live paths) |
| #72 | refactor | Gate logic extracted to `evaluate-gates.sh`; independently testable |
| #73 | fix | `/comprehensive-review-help` skill replaced dead placeholder with HELP.md output |
| #74 | feat | bats test suite: fixtures, `test_helper.bash`, `load_function()` helper |
| #75 | feat | Golden orchestration contract tests (24 assertions) |
| #76 | feat | `category` field added to `json-findings` contract with 11-value taxonomy |
| #77 | feat | Novelty pass annotates recurring Low/Medium findings from prior reviews |
| #78 | feat | Auto-cheap routing: DOCS_ONLY and LOW_RISK_CONFIG modes skip Opus agents |

## Test plan

- [x] `bats tests/*.bats` — 54/54 tests pass (verified locally)
- [x] `plugin.json` bumped to 1.9.0
- [x] CHANGELOG entry written
- [x] README updated: flags table, json-findings example, Contributing section
- [x] CLAUDE.md updated: corrects "no test suite" claim
- [ ] Run `/comprehensive-review` against this branch (skipped per user request — user will review manually)
- [ ] Tag `v1.9.0` and update marketplace entry after merge

## Files changed

- `skills/comprehensive-review/SKILL.md` — gate delegation, DOCS_ONLY/LOW_RISK_CONFIG routing, novelty pass (Phase 2f), step renumbering (2g for redaction), Phase 5 reporting
- `skills/comprehensive-review/HELP.md` — DOCS_ONLY/LOW_RISK_CONFIG auto-cheap section
- `skills/comprehensive-review/SEVERITY.md` — json-findings field contract with category taxonomy
- `skills/comprehensive-review/PROVIDERS.md` — gh api inline review, body field for all providers
- `skills/comprehensive-review/scripts/evaluate-gates.sh` — new standalone gate evaluation script
- `skills/comprehensive-review/scripts/run-cve-check.sh` — two-pass awk fix, category field in output
- `skills/comprehensive-review-help/SKILL.md` — functional help skill
- `agents/{architecture,security,blind-hunter,edge-case-hunter,adversarial-general}.md` — category field in json-findings
- `tests/` — new directory with 4 test files, `test_helper.bash`, and fixture directories
- `.claude-plugin/plugin.json` — version 1.9.0
- `CHANGELOG.md`, `README.md`, `CLAUDE.md` — release documentation

### CODE REVIEW

File: skills/comprehensive-review/scripts/evaluate-gates.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  echo "GATE_CONTROL_FLOW=true"
+  echo "GATE_SECURITY_PATTERNS=true"
+  echo "GATE_CODE_OR_INFRA=true"
+  exit 0
+fi
+
+# Gate: has_error_patterns — fires silent-failure-hunter
+GATE_ERROR_PATTERNS=false
+grep -qE 'catch\b|if err|try \{|rescue\b|Result<|unwrap\b|\.error\(|\.expect\(|runCatching|guard\b|throws\b' "$DIFF_FILE" \
+  && GATE_ERROR_PATTERNS=true
```
[the remark this answers is not in the document]
Comment (gchaix) [reply skills/comprehensive-review/scripts/evaluate-gates.sh]: Fixed in commit 2a65c8a: added `|| true` to the bare `grep && GATE_ERROR_PATTERNS=true` assignment.

File: skills/comprehensive-review/scripts/evaluate-gates.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+# Gate: has_control_flow — fires edge-case-hunter (added lines only via + prefix filter)
+GATE_CONTROL_FLOW=false
+grep -E '^\+' "$DIFF_FILE" | grep -qE '\b(if|elif|else|for|while|do|case|switch|match|try|catch|except|rescue|unless|when|loop|break|continue|return|goto|defer|finally)\b' \
+  && GATE_CONTROL_FLOW=true
```
[the remark this answers is not in the document]
Comment (gchaix) [reply skills/comprehensive-review/scripts/evaluate-gates.sh]: Fixed in commit 2a65c8a: added `|| true` to the GATE_CONTROL_FLOW grep pipeline.

File: tests/test_helper.bash [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+# Extract a single function definition from a script and eval it into the
+# current shell, so it can be tested in isolation without sourcing the whole
+# script (which would trigger the orchestration `set -euo pipefail` main loop).
+#
+# Brace-depth tracking ignores braces inside single-quoted strings (the awk
+# bodies in these scripts are single-quoted), which is sufficient for the
+# functions under test. It does NOT handle braces inside double-quoted strings
+# or heredocs — keep extracted functions free of those.
+#
```
[the remark this answers is not in the document]
Comment (gchaix) [reply tests/test_helper.bash:23]: Updated the comment in 2a65c8a to accurately describe the constraint: the brace-depth tracker handles single-quoted awk bodies correctly because bash awk programs like `awk '...' "$file"` have their `{ }` characters at the top level of the bash single-quoted string, not inside it. The stated constraint (no braces in bash double-quoted strings or heredocs) is accurate — those would miscount. All 55 tests pass with the current implementation, including parse_go_mod which has a multi-page awk body.

File: tests/test_helper.bash [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      }
+      body = body $0 "\n"
+      if (started && depth == 0 && body != "") { print body; exit }
+    }
+  ' "$script")
+
+  if [[ -z "$func_body" ]]; then
+    echo "ERROR: Could not extract function '${func_name}' from ${script}" >&2
+    return 1
+  fi
```
[the remark this answers is not in the document]
Comment (gchaix) [reply tests/test_helper.bash:47]: Fixed in commit 2a65c8a: replaced bare `eval "$func_body"` with `if ! eval "$func_body"; then echo ERROR ... >&2; return 1; fi`.

File: tests/run_cve_check.bats [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  cp "$CVE_FIX/go.mod.replace-before-require" "$WORK/go.mod"
+  OSV_MOCK_FILE="$CVE_FIX/osv-batch-malformed.json" run --separate-stderr "$SCRIPT" "$WORK/go.mod"
+  [ "$status" -eq 0 ]
+  [ "$output" = "[]" ]
+}
+
+@test "cve-check: CVSS v4 vector maps to High (conservative fallback)" {
+  cp "$CVE_FIX/go.mod.replace-before-require" "$WORK/go.mod"
+  OSV_MOCK_FILE="$CVE_FIX/osv-batch-cvss4.json" run --separate-stderr "$SCRIPT" "$WORK/go.mod"
+  [ "$status" -eq 0 ]
```
[the remark this answers is not in the document]
Comment (gchaix) [reply tests/run_cve_check.bats:122]: Fixed in commit 2a65c8a: added `echo "$output" | jq -e '.[0].category == "dependency-cve"' >/dev/null` to the CVSS v4 test.

File: tests/run_trufflehog.bats [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  echo "$output" | jq -e '.[0].severity == "High"' >/dev/null
+  echo "$output" | jq -e '.[0].confidence == 85' >/dev/null
+  echo "$output" | jq -e '.[0].file == "src/notifications.ts"' >/dev/null
+}
+
+@test "trufflehog: empty mock file returns empty array" {
+  TRUFFLEHOG_MOCK_FILE="/dev/null" run --separate-stderr "$SCRIPT" ""
+  [ "$status" -eq 0 ]
+  [ "$output" = "[]" ]
+}
```
[the remark this answers is not in the document]
Comment (gchaix) [reply tests/run_trufflehog.bats:51]: Fixed in commit 2a65c8a: added `unset TRUFFLEHOG_MOCK_FILE` as the first line of the test body.

File: skills/comprehensive-review/scripts/run-cve-check.sh [hunk 7]
```diff
@@ -301,44 +302,54 @@ osv_to_findings() {
...
               replaces[old_name] = new_name SUBSEP new_ver
           }
         }
-        next
       }
+      next
+    }
+
+    # ---- Pass 2 (second read): emit require rows with the complete replace map ----
+    # Reset state at the start of pass 2 in case pass 1 ended inside an unclosed block.
```
[the remark this answers is not in the document]
Comment (gchaix) [reply skills/comprehensive-review/scripts/run-cve-check.sh:343]: The concern is valid in isolation, but the guard is correct here because awk is invoked as `awk '...' "$file" "$file"` (line 370) — the file is passed twice on the command line. When awk starts reading the second argument, FNR resets to 1 while NR continues counting, so the condition `FNR==1 && NR!=1` correctly identifies the pass-2 boundary. That said, `NR > 1` expresses the intent more clearly and is equivalent. Changed to `NR > 1` in commit 613efe6 and added a comment explaining the invariant. Regression test (go.mod.unclosed-replace) still passes.

File: tests/run_cve_check.bats [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  echo "$output" | jq -e 'length > 0' >/dev/null
+  echo "$output" | jq -e '.[0].severity == "High"' >/dev/null
+  echo "$output" | jq -e '.[0].category == "dependency-cve"' >/dev/null
```
[the remark this answers is not in the document]
Comment (gchaix) [reply tests/run_cve_check.bats]: Fixed in commit c21b968: added `|| { echo "Actual category: $(echo "$output" | jq -r '.[0].category')" >&2; false; }` fallback to the assertion.

File: skills/comprehensive-review/scripts/evaluate-gates.sh [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+# Gate: has_control_flow — fires edge-case-hunter (added lines only via + prefix filter)
+# The first grep (extract + lines) may legitimately exit 1 (no added lines); treat that as false.
+# The second grep exits 2 only on I/O error, which we abort on.
+GATE_CONTROL_FLOW=false
+_added_lines=$(grep -E '^\+' "$DIFF_FILE" || true)
+if [[ -n "$_added_lines" ]]; then
+  _rc=0
+  echo "$_added_lines" | grep -qE '\b(if|elif|else|for|while|do|case|switch|match|try|catch|except|rescue|unless|when|loop|break|continue|return|goto|defer|finally)\b' || _rc=$?
+  if [[ $_rc -eq 2 ]]; then echo "ERROR: grep failed on added lines" >&2; exit 1; fi
+  if [[ $_rc -eq 0 ]]; then GATE_CONTROL_FLOW=true; fi
```
[the remark this answers is not in the document]
Comment (gchaix) [reply skills/comprehensive-review/scripts/evaluate-gates.sh:54]: Fixed in commit 9269273: replaced `|| true` with explicit `_rc` capture — exit 2 now aborts with an error message, consistent with the second grep in the same block.
